### PR TITLE
LDP/LDR/STR instruction translation

### PR DESCRIPTION
**Pull request by anthonymckay**

### CODE REVIEW

File: lib/Target/AArch64/DC/AArch64DCInstruction.cpp [hunk 1]
```diff
@@ -93,16 +93,19 @@ Value *AArch64DCInstruction::translateCustomOperand(unsigned OperandType,
...
+  case AArch64::OpTypes::am_bl_target:
+  case AArch64::OpTypes::am_ldrlit: {
     auto *ResTy = Builder.getInt8PtrTy();
 
-    // bl target is an offset in number of (4byte) instructions from PC
+    // target is an offset in number of (4byte) instructions from PC
     // target = PC + (imm * 4)
     // TODO: add check that offset is +-128MB from PC?
+    //       aml_ldrlit is identical except +-1MB from PC.
     signed offset = getImmOp(MIOperandNo)*4;
```
Comment (repzret) [inline lib/Target/AArch64/DC/AArch64DCInstruction.cpp]: Heh, "signed" surprised me.  Use the more common "int" instead?

Also, naming convention, here and below ("Offset", "ImmTarget").

File: lib/Target/AArch64/AArch64Sema.td [hunk 2]
```diff
@@ -27,30 +27,84 @@ def : Semantics<BLR, [(set LR, PC), (AArch64call GPR64:$Rn)]>;
...
+
   def : Semantics<!cast<Instruction>("STP" # regname # "pre"), [
           (store (Ty regtype:$Rt), (bitconvert (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset)))),
           (store (Ty regtype:$Rt2), (bitconvert (add scale, (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset))))),
           (set GPR64sp:$wback, (bitconvert (add scale, (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset)))))]>;
 
+  def : Semantics<!cast<Instruction>("LDP" # regname # "pre"), [
+          (set (Ty regtype:$Rt), (Ty (load (bitconvert (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset)))))),
+          (set (Ty regtype:$Rt2), (Ty (load (bitconvert (add scale, (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset))))))),
+          (set GPR64sp:$wback, (bitconvert (add scale, (add GPR64sp:$Rn, (sext !cast<Operand>("simm7s" # scale):$offset)))))]>;
```
Comment (repzret) [inline lib/Target/AArch64/AArch64Sema.td]: I don't think the "add scale" is correct.  Isn't the only added value the offset?
Also, are these wback "bitconvert"s necessary?

(I just noticed that STP has the same problems).

Comment (anthonymckay) [inline lib/Target/AArch64/AArch64Sema.td]: the "add scale" part is to add the operand width (4 bytes, 8 bytes, etc) to the destination memory address (or source memory addr in the case of LDP), the scale value just happens to correspond with that. e.g. for the X reg STP:
```
dst+offset = src1
dst+offset+8 = src2
```

Comment (repzret) [inline lib/Target/AArch64/AArch64Sema.td]: Sure, that's fine for the load/store pointers.  But I'm worried about the writeback value: IIRC it doesn't take into account the scale, only the offset.

**Pull request merged**